File: golang/d3/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	part1()
	part2()
}
func part2() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Printf("error opening the file: %v", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var counter uint

	for scanner.Scan() {
		line1 := scanner.Text()
		scanner.Scan()
		line2 := scanner.Text()
		scanner.Scan()
		line3 := scanner.Text()
		badgeItem := findTrip(line1, line2, line3)
		counter += runeToPriority(badgeItem)
	}
	println(counter)
}

func findTrip(ruck1, ruck2, ruck3 string) rune {
	for _, item := range ruck1 {
		if strings.ContainsRune(ruck2, item) && strings.ContainsRune(ruck3, item) {
			return item
		}
	}
	return rune(0)
}

func part1() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Printf("error opening the file: %v", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var counter uint

	for scanner.Scan() {
		line := scanner.Text()
		middle := len(line) / 2
		dup := findDup(line[:middle], line[middle:])
		counter += runeToPriority(dup)
	}
	println(counter)
}

func findDup(comp1, comp2 string) rune {
	for _, item := range comp1 {
		if strings.ContainsRune(comp2, item) {
			return item
		}
	}
	return rune(0)
}

func runeToPriority(r rune) uint {
	if 'a' <= r && r <= 'z' {
		return uint(r - 'a' + 1)
	} else if 'A' <= r && r <= 'Z' {
		return uint(r - 'A' + 27)
	}
	return 0
}

File: rust/d3/src/main.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

fn main() {
    part1();
    part2();
}

fn part2() {
    let file = File::open("input.txt").expect("Fail to open file");
    let reader = BufReader::new(file);
    let mut counter: u32 = 0;

    let mut lines = reader.lines().map(|line| line.expect("fail to read line"));

    while let Some(line1) = lines.next() {
        let line2 = lines.next().expect("Expected another line");
        let line3 = lines.next().expect("Expected another line");

        let badge_item = find_trip(&line1, &line2, &line3);
        counter += char_to_priority(badge_item);
    }
    println!("{counter}")
}

fn find_trip(ruck1: &str, ruck2: &str, ruck3: &str) -> char {
    for ch in ruck1.chars() {
        if ruck2.contains(ch) && ruck3.contains(ch) {
            return ch;
        }
    }
    '\0'
}

fn part1() {
    let file = File::open("input.txt").expect("Fail to open file");
    let reader = BufReader::new(file);
    let mut counter: u32 = 0;

    for line in reader.lines() {
        let l = line.expect("Fail to read line");
        let middle = l.len() / 2;
        if let Some(dup_val) = find_dup(&l[..middle], &l[middle..]) {
            counter += char_to_priority(dup_val);
        }
    }
    println!("{counter}")
}

fn find_dup(comp1: &str, comp2: &str) -> Option<char> {
    comp1.chars().find(|&item| comp2.contains(item))
}

fn char_to_priority(c: char) -> u32 {
    match c {
        'a'..='z' => c as u32 - 'a' as u32 + 1,
        'A'..='Z' => c as u32 - 'A' as u32 + 27,
        _ => 0,
    }
}
